pyface/fields/tests/test_combo_field.py

```diff
@@ -1,163 +1,114 @@
+        with self.assertTraitChanges(self.widget, 'value', count=1):
+            self.widget._set_control_value(2)
+            self.gui.process_events()
 
-            self.assertEqual(widget.value, 2)
-        finally:
-            widget.destroy()
+        self.assertEqual(self.widget.value, 2)
 
     def test_combo_field_icon_formatter(self):
         image = ImageResource('question')
-        widget = ComboField(
-            parent=None,
-            values=[1, 2, 3, 4],
-            tooltip='Dummy',
-            formatter=lambda x: (image, str(x)),
-        )
+        self.widget.values = [1, 2, 3, 4]
+        self.widget.formatter = lambda x: (image, str(x))
+        self._create_widget_control()
 
-        widget._create()
-        try:
-            widget.show(True)
-            widget.value = 2
+        self.widget.value = 2
+        self.gui.process_events()
 
-            self.assertEqual(widget._get_control_value(), 2)
-            self.assertEqual(widget._get_control_text(), '2')
-        finally:
-            widget.destroy()
+        self.assertEqual(self.widget._get_control_value(), 2)
+        self.assertEqual(self.widget._get_control_text(), '2')
 
     def test_combo_field_values(self):
-        widget = ComboField(
-            parent=None,
-            values=['one', 'two', 'three', 'four'],
-            tooltip='Dummy',
-        )
-
-        widget._create()
-        try:
-            widget.show(True)
-            widget.values = ['four', 'five', 'one', 'six']
-
-            self.assertEqual(widget.value, 'four')
-        finally:
-            widget.destroy()
-
-    def test_combo_field_tooltip(self):
-        widget = ComboField(
-            parent=None,
-            values=['one', 'two', 'three', 'four'],
-            tooltip='Dummy',
-        )
-
-        widget._create()
-        try:
-            widget.show(True)
-
-            widget.tooltip = "New tooltip."
-
-            self.assertEqual(widget._get_control_tooltip(), "New tooltip.")
-        finally:
-            widget.destroy()
-
-    def test_combo_field_menu(self):
-        widget = ComboField(
-            parent=None,
-            values=['one', 'two', 'three', 'four'],
-            tooltip='Dummy',
-        )
-
-        widget._create()
-        try:
-            widget.show(True)
-            widget.menu = MenuManager(Action(name='Test'), name='Test')
-        finally:
-            widget.destroy()
+        self._create_widget_control()
+
+        self.widget.values = ['four', 'five', 'one', 'six']
+        self.gui.process_events()
+
+        # XXX different results in Wx and Qt
+        # As best I can tell, difference is at the Traits level
+        # On Qt setting 'values' sets 'value' to "four" before combofield
+        # handler sees it.  On Wx it remains  as "one" at point of handler
+        # call.  Possibly down to dictionary ordering or something
+        # similar.
+        self.assertIn(self.widget.value, {'one', 'four'})
```

Comment:
Is this non-deterministic?

Reply:
No, it seems to always be one way in the Qt implementation and the other for Wx but the difference occurs _prior_ to the toolkit code being called (ie. on one backend the `_values_updated` handler gets called with `self.value == 'one'` and on the other it gets called with `self.value == 'four'`.  So my suspicion is that there is a trait handler to reset the `value` trait on changes to `values` and it is not specified which one gets called first.

So my fear is that changing the structure of each trait class, or running in a different context, may case non-deterministic behaviour; but I haven't dug deeper, because even if I do find out what is causing it, I suspect there is nothing much I can _do_ about it.

Reply:
OK. I suppose this note will have to suffice for the time being.